2020/9/encoding/src/main.rs

```diff
@@ -0,0 +1,105 @@
+use std::env;
+use std::fs::File;
+use std::io::{BufRead, BufReader, Error};
+
+fn main() -> Result<(), Error> {
+    let args: Vec<String> = env::args().collect();
+    let reader = BufReader::new(File::open(&args[1])?);
+
+    let input : Vec<_> = reader.lines().map(|line| line.unwrap().parse::<usize>().unwrap()).collect();
+    if let Some(weakness) = encryption_weakness(&input, 25) {
+        println!("Found encryption weakness: {}", weakness);
+    } else {
+        println!("Didn't find an encryption weakness.");
+    }
+
+    Ok(())
+}
+
+pub fn is_last_valid(num: usize, input: &[usize]) -> bool {
+    for x in 0..input.len() - 1 {
+        for y in x + 1..input.len() {
+            //if num == 863 {
+            //    println!("Checking {} + {} = {},", input[x], input[y], input[x] + input[y]);
+            //}
+            if input[x] + input[y] == num && input[x] != input[y] {
+                return true;
+            }
+        }
+    }
+
+    false
+}
+
+pub fn first_invalid(input: &[usize], preamble_length: usize) -> Option<(usize, usize)> {
+    // Consider - storing a set of previous sums, along with the indices of their addends would allow
+    // fewer iterations if the input was super long, or the preamble length was longer.
+    for x in preamble_length..input.len() {
+        //println!("looking at index {} from {} to {}", x, x-preamble_length, x);
+        if !is_last_valid(input[x], &input[x-preamble_length..x]) {
+            return Some((x,input[x]));
+        }
+    }
+
+    None
+}
+
+pub fn encryption_weakness(input: &[usize], preamble_size: usize) -> Option<usize> {
+    let (index, first_invalid) = first_invalid(input, preamble_size)?;
+
+    // Let's just brute force this - how many subsets can there be?
```

Comment:
Turns out this is the first time this year it has taken a perceptible amount of time to solve one of these 🤣. About half a second.

Reply:
Just wait until Day 10 😉

Reply:
Yeah, that's basically what I said to my wife last night.